text/0016-internationalization.md

```diff
@@ -0,0 +1,117 @@
+
+  
+
+Transifex is a cloud-based localization platform built to help you manage the translation and localization of your app, website, video subtitles, and more. It acts as a repository for your content (think GitHub for translation) and includes tools for developers to get that content into Transifex automatically. Transifex also provides translators a web interface to submit translations.
+
+  
+
+## Decision, including impact on distributions
+
+OpenMRS is implemented throughout the world and is used by speakers of many different languages. As not every user speaks English as a primary language, we will:
+
+1.  Provide an ESM module to translate content in the UI to the user’s specified language.
+    
+2.  The default language is English, this should be made visible and explicit in the code.
+    
+3.  The ESM translation module will be able to translate content **without** connectivity.
+    
+4.  Presently, OpenMRS uses Transifex ([https://www.transifex.com/openmrs/OpenMRS/dashboard/](https://www.transifex.com/openmrs/OpenMRS/dashboard/)) to manage translations. We will continue to use Transifex to support translations for the MF project.
+    
+5.  Possibly find ways to reuse OpenMRS module translations found in the `message.properties` file.
+
+7.  Update mechanism: Once translations are marked as complete, create a mechanism to publish the translated strings to a github repo. 
+    
+
+## Reason for decision
+
+- Most distributions are deployed to areas with no to intermittent connectivity, thus the translations module should be completely offline.
+
+- English is the lingua franca and spoken in most parts of the world.
+
+- Reasons for choosing Transifex:
+	 - Transifex is the most popular internationalization SaaS and offers tools that would make the translations workflow smoother.
+	 - The community is already familiar using transifex for translation services.
+
+
+## Alternatives
+- No translation service at all and default to English for all frontend modules.
+-  Use another service provider built specifically for frontend applications such as [react-i18next](https://github.com/i18next/react-i18next)
+- We could store the translations with each `esm` module instead of having a translations `esm` module.
+- Repo to handle translations service: we could have the translations handled by `openmrs-esm-api`
+
+
+## Common practices (not enforced)
+
+
+### Definitions
+
+Messages - OpenMRS historically used this word to refer to the content that would be replaced.
+
+  
+### Structure of Message File
+
+Here’s an example of how message.properties file looks like:
+
+[https://github.com/openmrs/openmrs-module-legacyui/blob/master/api/src/main/resources/messages.properties](https://github.com/openmrs/openmrs-module-legacyui/blob/master/api/src/main/resources/messages.properties)
```

Comment:
This file is a java-specific format (see [explanation](https://crunchify.com/java-properties-file-how-to-read-config-properties-values-in-java/)).

I'd prefer us to use a different format that is more friendly to javascript, such as JSON, for the frontend translations. I think it would be okay to parse and use `.properties` files if needed to interop with existing things (could use something like [this npm package](https://www.npmjs.com/package/properties-file) to pull it off), but I don't think it makes sense to put `.properties` files into our frontend code if we don't need to.

Reply:
I agree. 👍